File: packages/sdk-elixir/lib/zenvio/messages.ex
defmodule Zenvio.Messages do
  @moduledoc """
  Messages (templates) — POST /v1/templates/send
  Envio por template em múltiplos canais (whatsapp, sms, email).
  Parâmetros (mesmos da API): to, template, variables, channels, instance_id, from, fromName.
  """

  @doc """
  POST /v1/templates/send.
  params: map com "to", "template", "channels" (obrigatórios) e opcionalmente "variables", "instance_id", "from", "fromName".
  Aceita também "from_name" (será enviado como "fromName" para a API).
  """
  def send(client, params, opts \\ []) do
    params = normalize_params(params)
    Zenvio.request(client, :post, "/templates/send", params, opts)
  end

  defp normalize_params(params) when is_map(params) do
    params
    |> maybe_put_from_name()
  end

  defp maybe_put_from_name(params) do
    from_name = Map.get(params, "from_name") || Map.get(params, :from_name)
    from_name_api = Map.get(params, "fromName") || Map.get(params, :fromName)
    cond do
      is_binary(from_name) and is_nil(from_name_api) ->
        params
        |> Map.delete("from_name")
        |> Map.delete(:from_name)
        |> Map.put("fromName", from_name)
      true ->
        params
    end
  end
end

File: packages/sdk-php/src/Messages.php
<?php

namespace Zenvio;

/**
 * Messages (templates) — POST /v1/templates/send
 */
class Messages
{
    private Zenvio $client;

    public function __construct(Zenvio $client)
    {
        $this->client = $client;
    }

    /**
     * POST /v1/templates/send — envio por template (whatsapp, sms, email).
     * Parâmetros (mesmos da API): to, template, variables?, channels, instance_id?, from?, fromName?
     *
     * @param array{to: list<string>, template: string, variables?: array<string, string|int>, channels: list<string>, instance_id?: string, from?: string, fromName?: string} $params
     */
    public function send(array $params): array
    {
        if (isset($params['from_name']) && !\array_key_exists('fromName', $params)) {
            $params['fromName'] = $params['from_name'];
            unset($params['from_name']);
        }
        return $this->client->request('POST', '/templates/send', $params);
    }
}
